osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs

```diff
@@ -133,13 +133,23 @@ private double locallyCombinedDifficulty(Colour colour, Rhythm rhythm, Stamina s
         {
             List<double> peaks = new List<double>();
 
-            for (int i = 0; i < colour.StrainPeaks.Count; i++)
+            var colourPeaks = colour.CurrentStrainPeaks.GetEnumerator();
+            var rhythmPeaks = rhythm.CurrentStrainPeaks.GetEnumerator();
+            var staminaRightPeaks = staminaRight.CurrentStrainPeaks.GetEnumerator();
+            var staminaLeftPeaks = staminaLeft.CurrentStrainPeaks.GetEnumerator();
+
+            do
             {
-                double colourPeak = colour.StrainPeaks[i] * colour_skill_multiplier;
-                double rhythmPeak = rhythm.StrainPeaks[i] * rhythm_skill_multiplier;
-                double staminaPeak = (staminaRight.StrainPeaks[i] + staminaLeft.StrainPeaks[i]) * stamina_skill_multiplier * staminaPenalty;
+                double colourPeak = colourPeaks.Current * colour_skill_multiplier;
+                double rhythmPeak = rhythmPeaks.Current * rhythm_skill_multiplier;
+                double staminaPeak = (staminaRightPeaks.Current + staminaLeftPeaks.Current) * stamina_skill_multiplier * staminaPenalty;
                 peaks.Add(norm(2, colourPeak, rhythmPeak, staminaPeak));
-            }
+            } while (colourPeaks.MoveNext() && rhythmPeaks.MoveNext() && staminaRightPeaks.MoveNext() && staminaLeftPeaks.MoveNext());
+
+            colourPeaks.Dispose();
+            rhythmPeaks.Dispose();
+            staminaRightPeaks.Dispose();
+            staminaLeftPeaks.Dispose();
```

Comment:
This code is too ugly to live. I did not anticipate that changing the peaks to `IEnumerable` would cause this, but in light of it I'd rather have a few redundant list copies in order to be able to access by index normally.

Either that, or do something with `.Zip()`, but I think any solution that would use that is likely to be too ugly as well.

Reply:
I agree it's not as elegant as indexed for loop, or foreach with `Zip()` but I certainly don't agree that it's "too ugly to live'...
By all rights, this is a textbook intended usage of multiple low level enumerators.
I've cleaned it up slightly in 05278e0510ef2b0492c41296978cb585811e7901 by using `using` blocks to avoid the `Dispose()` calls.
I'd be keen to get other's thought on this, because I personally don't think this is a good enough reason to justify intentional efficiency impacts (no matter how small) on an efficiency sensitive part of the codebase.

Reply:
If performance is a worry then the only thing that will dispel any and all doubts is a benchmark, and not theory crafting.

Reply:
My point is that the reason we started down this path in the first place was because you didn't want the needless copy https://github.com/ppy/osu/pull/11689#discussion_r571820236.
I'm still of the opinion that the best method is either what I originally did https://github.com/ppy/osu/pull/11689#discussion_r571817264, or the enumerable approach, since a copy in the property is an unexpected allocation. There's no need to benchmark, since the difference is literally sub-millisecond, but its about the principle.
Regardless, I'm tired of arguing implementation details. I've changed it to have a copy in the property.

Changed in b6ac32bddd5bc05a215b3710ea1dbfcf44ba1a89

Reply:
For the record: The primary reason I requested this change is that using raw enumerators is rather rare, using four at a time like the previous code did - especially so. I think making some list copies is a small price to pay in exchange for adhering to the principle of least astonishment. Feel free to overrule me on this, but I don't think the impact of the list copies is large enough to justify the code being non-obvious to comprehend.

Reply:
> using raw enumerators is rather rare, using four at a time like the previous code did - especially so

Rare circumstances require rare techniques.

While I do disagree with your opinion here, given that you have already gone ahead and committed further changes on top of this by yourself, I don't think its worth the effort to push the issue any further.

I'm happy to leave it as you have committed, however I do stand by my opinion that what I originally did in 3743937774fb21887ee2472555274a5784d44048 is the most elegant solution; also adhering to the principle of least astonishment.